airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/ConfiguredAirbyteCatalogMigrator.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.protocol;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.protocol.migrations.ConfiguredAirbyteCatalogMigration;
+import io.airbyte.commons.protocol.migrations.MigrationContainer;
+import io.airbyte.commons.version.Version;
+import jakarta.annotation.PostConstruct;
+import jakarta.inject.Singleton;
+import java.util.List;
+import java.util.Set;
+
+@Singleton
+public class ConfiguredAirbyteCatalogMigrator {
+
+  private final MigrationContainer<ConfiguredAirbyteCatalogMigration<?, ?>> migrationContainer;
+
+  public ConfiguredAirbyteCatalogMigrator(final List<ConfiguredAirbyteCatalogMigration<?, ?>> migrations) {
+    migrationContainer = new MigrationContainer<>(migrations);
+  }
+
+  @PostConstruct
+  public void initialize() {
+    migrationContainer.initialize();
+  }
+
+  /**
+   * Downgrade a message from the most recent version to the target version by chaining all the
+   * required migrations
+   */
+  public <PreviousVersion, CurrentVersion> PreviousVersion downgrade(final CurrentVersion message, final Version target) {
+    return migrationContainer.downgrade(message, target, ConfiguredAirbyteCatalogMigrator::applyDowngrade);
+  }
```

Comment:
Do we need to validate that this version is indeed a downgrade?

Reply:
I don't think so, current design enforces a one-way look up from/to the current version. If the version provided isn't a valid downgrade or upgrade, we either fail because the version isn't supported (in the list of known migrations) or a mismatch in the type conversion.